### PR TITLE
Support reflecting `CXXFLAGS` and `LDFLAGS` environment variables

### PR DESCRIPTION
**Pull request by onion108**
Added support for environment variables `CXXFLAGS` and `LDFLAGS` while generating makefile.

See https://github.com/poac-dev/poac/issues/923.

### CODE REVIEW

**Comment by ken-matsui** [comment]
Why would you prefer to do this instead of [my idea](https://github.com/poac-dev/poac/issues/923#issuecomment-2113612533)?

**Comment by onion108** [comment]
Firstly, the most important reason is, as far as I know, many build systems that are widely used made use of those environment variables (`autoconf`, Makefile, CMake etc.). I think it will be less confusing to make the behavior consistent as those build system. (Even `cargo` respects `RUSTFLAGS`)

Also sometimes I would like to temporarily change the flags frequently to quickly see the result and adjust my flags in the config after I'm confident about the result. I personally prefer setting the environment variables to edit the config file over and over again because it's convenient to just add an environment variable in the command line, and I don't need to repeatly edit and save a file just for something experimental (in my personal opinion).

Another reason (may be too personal) is that, if I downloaded a project that I don't own and I want to build it with my preferred toolchain, for instance I want to build with Homebrew ruby instead of the default ruby provided by macOS (but the project's config doesn't contain flags for Homebrew ruby), it's just convenient to just set an environment variable instead of editing config file, because if I changed the project config file and next time I want to upgrade the repository, I still need to revert the change, pull the repository and change it again. It may fail the compilation or even have different behavior after the compilation but since those environment variables were set by me, it means I know the risks and I'll take them. ~(And actually I set it globally, in real life)~

By the way, (not very seriously though) we supported `CXX` environment variable so why not support others as well?

To conclude I think the most important reason is to keep the behavior consistent as build systems that are widely used, and it will be also super convenient for some users to quickly edit the flags just for experimental uses (especially command-line nerds like me, yeah). About the potential risks it may bring, like compilation error or different behavior etc., since the system won't set those environment variables automatically (and it means they are set only by user) I think user would and should take the risks, and once user set the environment variable (even globally) they usually knows what they're doing (and hoping the build system uses those variables).

**Comment by onion108** [comment]
P.S.: I do think your idea is great and should be used on most projects, but I think the support of environment flags still should be implemented as well for the consistency with other widely-used build systems and some advanced user may require it to work

**Comment by ken-matsui** [comment]
Thanks for the clarification, I'll review your PR when I get a chance. Please also make the PR title clearer.

File: src/BuildConfig.cc [hunk 1]
```diff
@@ -97,6 +97,68 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
...
+  if (!buffer.empty()) {
+    result.push_back(buffer);
+  }
+
+  return result;
+}
+
+static std::vector<std::string>
+getEnvFlags(const std::string& name) {
+  if (const auto* cenv = std::getenv(name.c_str())) {
```
Comment (ken-matsui) [inline src/BuildConfig.cc:154]: Please reduce the use of auto unless it's a clear redundancy.

File: src/BuildConfig.cc [hunk 2]
```diff
@@ -541,6 +603,10 @@ BuildConfig::setVariables() {
   for (const std::string_view flag : profile.cxxflags) {
     cxxflags.emplace_back(flag);
   }
+  for (const std::string& flag : getEnvFlags("CXXFLAGS")) {
+    cxxflags.emplace_back(flag);
+    logger::debug("adding env: ", flag);
```
Comment (ken-matsui) [inline src/BuildConfig.cc:608]: Do we need this?

File: src/BuildConfig.cc [hunk 3]
```diff
@@ -541,6 +603,10 @@ BuildConfig::setVariables() {
   for (const std::string_view flag : profile.cxxflags) {
     cxxflags.emplace_back(flag);
   }
+  for (const std::string& flag : getEnvFlags("CXXFLAGS")) {
```
Comment (ken-matsui) [inline src/BuildConfig.cc:606]: Let's add a comment of why `CXXFLAGS` will be appended at last; it's because environment variables take the highest precedence, right? (please also add a similar comment right above `LDFLAGS` as well)

File: src/BuildConfig.cc [hunk 4]
```diff
@@ -97,6 +97,68 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
...
+  }
+
+  return result;
+}
+
+static std::vector<std::string>
+getEnvFlags(const std::string& name) {
+  if (const auto* cenv = std::getenv(name.c_str())) {
+    std::string env(cenv);
+    std::vector<std::string> result;
```
Comment (ken-matsui) [inline src/BuildConfig.cc:156]: Are we using this?

File: src/BuildConfig.cc [hunk 5]
```diff
@@ -97,6 +97,68 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
   }
 }
 
+static std::vector<std::string>
+parseEnvFlags(const std::string& env) {
```
Comment (ken-matsui) [inline src/BuildConfig.cc:101]: This function is hard to follow; please make this clearer and add more comments.

File: src/BuildConfig.cc [hunk 6]
```diff
@@ -97,6 +97,68 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
...
+    result.push_back(buffer);
+  }
+
+  return result;
+}
+
+static std::vector<std::string>
+getEnvFlags(const std::string& name) {
+  if (const auto* cenv = std::getenv(name.c_str())) {
+    std::string env(cenv);
```
Comment (ken-matsui) [inline src/BuildConfig.cc:155]: Do we need this specific instantiation?

File: src/BuildConfig.cc (hunk 5 above)
Comment (ken-matsui) [inline src/BuildConfig.cc:101]: Since we just use the return value of `std::getenv` and then iterate over this, I think `std::string_view` is more proper.

**Review by ken-matsui — COMMENTED** [review]
Thank you for working on this. This is a good start; I left some comments.

File: src/BuildConfig.cc (hunk 4 above)
Comment (onion108) [reply src/BuildConfig.cc]: Cleaned up in the latest commit.

File: src/BuildConfig.cc (hunk 3 above)
Comment (onion108) [reply src/BuildConfig.cc:640]: Fixed in the latest commit.

File: src/BuildConfig.cc (hunk 2 above)
Comment (onion108) [reply src/BuildConfig.cc]: Cleaned up in the latest commit. (Just for testing when I'm playing around with the code and I forgot to remove it. I wrote a test for `parseEnvFlags` in the latest commit so now I can test it via `poac test`)

File: src/BuildConfig.cc (hunk 5 above)
Comment (onion108) [reply src/BuildConfig.cc]: Actually `std::string_view` can't be used there. 

Although this implementation is really just iterating over the string and seems to be possible just to take slices of it, it is actually because the makefile generator seems not escaping any character including spaces so I decided to reserve the backslashes and quotes temporarily, but they should be removed in fact, something like `\ ` will be replaced with ` ` and `abc"def"ghi` will become `abcdefghi`, etc.

I just thought about this idea again and decided remove the backslashes and quotes after parsing and properly interpret them because reserving them may cause problems when generating compilation database. I don't know if there's an issue about that the makefile generator doesn't escape things correctly but I think there should be an issue or a fix about it (maybe I'll create the issue if it doesn't exist later, if I finished working on the PR). Not going to fix that (I mean the problem that makefile generator doesn't escape properly) in this PR because it's unrelated.

Comment (onion108) [reply src/BuildConfig.cc]: Added detailed comments in the latest commit. Also changed the code a little bit for some edge cases.

File: src/BuildConfig.cc [hunk 7]
```diff
@@ -97,6 +97,99 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
...
+
+  // This flag will be set when quote encountered.
+  // The quote encountered will be stored in the ignoreNextSeqStart to
+  // distinguish different quote types.
+  bool ignoreNextSeq = false;
+  char ignoreNextSeqStart = ' ';
+
+  result.emplace_back();
+
+  // There are bunches of conditional predicates, and to avoid if-else chain the
```
Comment (ken-matsui) [inline src/BuildConfig.cc:119]: I think the if-else chain can be clearer (also more compact) than many `continue`'s.

File: src/BuildConfig.cc [hunk 8]
```diff
@@ -97,6 +97,99 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
...
+  // special cases will be processed first and end in `continue` to skip rest of
+  // the loop. Same in special cases that are in another special case.
+  for (char i : env) {
+
+    if (ignoreNext) {
+      result.back() += i;
```
Comment (ken-matsui) [inline src/BuildConfig.cc:125]: Let's use `std::string buffer`, and then this can be clearer. We also don't need to do `result.emplace_back()`.

**Review by ken-matsui — COMMENTED** [review]
Overall, this looks great. Thanks for your work. Let's tweak more details.

File: src/BuildConfig.cc [hunk 9]
```diff
@@ -97,6 +97,99 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
...
+parseEnvFlags(std::string_view env) {
+  std::vector<std::string> result;
+
+  // This flag will be set when a backslash encountered.
+  bool ignoreNext = false;
+
+  // This flag will be set when quote encountered.
+  // The quote encountered will be stored in the ignoreNextSeqStart to
+  // distinguish different quote types.
+  bool ignoreNextSeq = false;
```
[the remark this answers is not in the document]
Comment (ken-matsui) [reply src/BuildConfig.cc:114]: Yep!

**Comment by onion108** [comment]
All review comments are resolved in the latest commit.

By the way I enabled workflow on the fork repo so I could run poac tidy and see the results to prevent me from making dumb mistakes. For some strange reasons *(that I don't know)* `poac tidy`'s output in the actions is quite different from my machine *(`clang-tidy` 19)*. My tidy seems telling me about almost nothing about my mistakes *(like making `const` when possible)*, but always complains about circular header file dependency in Git2 library downloaded via Homebrew *(which doesn't exists in the workflow output)*.

Another thing about tidy is after I committed my latest change, `poac tidy` *(on the Github Workflow, hence the correct one)* told me that in my test the code
```cpp
assertEq(argsEscapeMixed.size(), 5);
                                ~^~
```
is a magic constant and tell me to consider replacing it with a named constant *(but tbh we really need a named constant like `ARGS_ESCAPE_MIXED_SIZE_EXPECTED` just to check if the size is correct ...? I'm not quite sure about that but definitely it will not pass the `clang-tidy` action if I don't do so)*.

After digging deeper in the project I found a file `Algos.cc` that I think may be better for `parseEnvFlags` and its tests *(and perhaps `getEnvFlags` as well?)* to be placed in. I may consider doing in the next commit *(after checking the code itself is fine and no dumb mistakes or naming mistakes or such)*, but I'm not quite sure if I'm going to move both `parseEnvFlags` and `getEnvFlags`, making `parseEnvFlags` internal to `Algos.cc` or just move `parseEnvFlags` to that file *(and declare it in the header file as well)*.

**Comment by ken-matsui** [comment]
> All review comments are resolved in the latest commit.

Thanks!

> 
> By the way I enabled workflow on the fork repo so I could run poac tidy and see the results to prevent me from making dumb mistakes. For some strange reasons _(that I don't know)_ `poac tidy`'s output in the actions is quite different from my machine _(`clang-tidy` 19)_. My tidy seems telling me about almost nothing about my mistakes _(like making `const` when possible)_, but always complains about circular header file dependency in Git2 library downloaded via Homebrew _(which doesn't exists in the workflow output)_.

This problem was due to the previous `poac tidy` implementation, and I fixed it. Currently, the fix is available only on the main. GitHub Actions one is always using the source code from the commit that triggered the action, which always contains my fix anyway, so that's why we see differences.

> 
> Another thing about tidy is after I committed my latest change, `poac tidy` _(on the Github Workflow, hence the correct one)_ told me that in my test the code
> 
> ```c++
> assertEq(argsEscapeMixed.size(), 5);
>                                 ~^~
> ```
> 
> is a magic constant and tell me to consider replacing it with a named constant _(but tbh we really need a named constant like `ARGS_ESCAPE_MIXED_SIZE_EXPECTED` just to check if the size is correct ...? I'm not quite sure about that but definitely it will not pass the `clang-tidy` action if I don't do so)_.

Using a named constant here makes no sense, so let's use `NOLINT` (or `NOLINTNEXTLINE` depending on the code elegance). See [here](https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics) for more info.

> 
> After digging deeper in the project I found a file `Algos.cc` that I think may be better for `parseEnvFlags` and its tests _(and perhaps `getEnvFlags` as well?)_ to be placed in. I may consider doing in the next commit _(after checking the code itself is fine and no dumb mistakes or naming mistakes or such)_, but I'm not quite sure if I'm going to move both `parseEnvFlags` and `getEnvFlags`, making `parseEnvFlags` internal to `Algos.cc` or just move `parseEnvFlags` to that file _(and declare it in the header file as well)_.

Currently, they are used only by `BuildConfig.cc` so it makes sense to put them here. There's no compelling reason to make them project public APIs. Thanks for the suggestion anyway :)

**Comment by onion108** [comment]
Thanks for your replay. Added `NOLINTNEXTLINE` for `*-magic-numbers` before size checks and checked Github Actions. All the tests on the Github Actions passed now.
スクリーンショット 2024-11-04 10 51 30

File: src/BuildConfig.cc [hunk 10]
```diff
@@ -97,6 +97,69 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
...
+// the quotes and some escape sequences. (More specifically it will ignore
+// whatever character that goes after a backslash and preserve all characters,
+// usually used to pass an argument containing spaces, between quotes.)
+static std::vector<std::string>
+parseEnvFlags(std::string_view env) {
+  std::vector<std::string> result;
+  std::string buffer;
+
+  bool foundBackslash = false;
+
```
Comment (ken-matsui) [inline src/BuildConfig.cc:110]: Let's remove this newline here.

File: src/BuildConfig.cc [hunk 11]
```diff
@@ -97,6 +97,69 @@ BuildConfig::BuildConfig(const std::string& packageName, const bool isDebug)
...
+      } else {
+        buffer += c;
+      }
+    } else if (c == '\'' || c == '"') {
+      isInQuote = true;
+      quoteChar = c;
+    } else if (c == '\\') {
+      foundBackslash = true;
+    } else if (std::isspace(c)) {
+      // Add argumnet only if necessary (i.e. buffer is not empty)
```
Comment (ken-matsui) [inline src/BuildConfig.cc:133]:
```suggestion
      // Add argument only if necessary (i.e. buffer is not empty)
```

File: src/BuildConfig.cc [hunk 12]
```diff
@@ -997,6 +1072,46 @@ testDependOnUnregisteredTarget() {
...
+      " \"-I/path/contains space\"  '-Lanother/path with/space' normal  "
+  );
+  // NOLINTNEXTLINE(*-magic-numbers)
+  assertEq(argsEscapeQuotes.size(), 3);
+  assertEq(argsEscapeQuotes[0], "-I/path/contains space");
+  assertEq(argsEscapeQuotes[1], "-Lanother/path with/space");
+  assertEq(argsEscapeQuotes[2], "normal");
+
+  std::vector<std::string> argsEscapeMixed = parseEnvFlags(
+      R"-( "-IMy \"Headers\"/v1" '\?pattern' normal path/contain/\"quote\" mixEverything" abc "\?\#   )-"
```
Comment (ken-matsui) [inline src/BuildConfig.cc:1102]: Let's also test double backslashes. (I know your code works with this case :)

**Review by ken-matsui — COMMENTED** [review]
Looks great. After tweaking small things, ok for main. Thanks.

**Review by ken-matsui — APPROVED** [review]
Thanks for your time to work on this!
